api/config.go

```diff
@@ -147,5 +157,53 @@ func LoadDefaultConfig(args []string) (config Config, err error) {
 	if err = flagSet.Parse(args[1:]); err != nil {
 		return config, err
 	}
+
+	if config.Flag3DTConfig != "" {
+		return readConfigFile(config.Flag3DTConfig, config)
+	}
 	return config, nil
 }
+
+func readConfigFile(configPath string, defaultConfig Config) (Config, error) {
+	// load config
+	configContent, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return defaultConfig, err
+	}
+
+	if err := validateConfigFile(configContent); err != nil {
+		return defaultConfig, err
+	}
+
+	// override default values
+	err = json.Unmarshal(configContent, &defaultConfig)
+	return defaultConfig, err
+}
+
+// returns nil if config is valid
+func validateConfigFile(configContent []byte) error {
+	configMap := make(map[string]interface{})
+	if err := json.Unmarshal(configContent, &configMap); err != nil {
+		return fmt.Errorf("Error validating config file: %s", err)
+	}
+
+	// validate every single key/value pair in config file
+	for k, v := range configMap {
+		configMapItem := make(map[string]interface{}, 1)
+		configMapItem[k] = v
+		configItem, err := json.Marshal(configMapItem)
+		if err != nil {
+			return fmt.Errorf("Error validating config field: %s. %s", k, err)
+		}
+
+		tmpConfig := Config{}
+		if err := json.Unmarshal(configItem, &tmpConfig); err != nil {
+			return fmt.Errorf("Error validating config field: %s. %s", k, err)
+		}
+
+		if reflect.DeepEqual(tmpConfig, Config{}) {
```

Comment:
the other poor-man's approach would be to make all your fields pointers. then you can more easily tell when they're present vs zero-values. if you don't want the rest of the dependent code to deal with reading (possibly nil) pointers then have two versions of config: DTOConfig (with pointers) and Config (without pointers) .. and conversion funcs to translate between the two.